include/sparrow/typed_array.hpp

```diff
@@ -0,0 +1,360 @@
+         *
+         * @pre The container must not be empty (\see empty()).
+         *
+         * @return A reference to the last element.
+         */
+        reference back();
+        const_reference back() const;
+        ///@}
+
+        // Iterators
+
+        ///@{
+        /* Returns an iterator to the first element.
+         * If the vector is empty, the returned iterator will be equal to end().
+         *
+         * @return An iterator to the first element.
+         */
+        iterator begin();
+        const_iterator begin() const;
+        const_iterator cbegin() const;
+        ///@}
+
+        ///@{
+        /**
+         * This element acts as a placeholder; attempting to access it results in undefined behavior.
+         *
+         * @return An iterator to the element following the last element of the vector.
+         */
+        iterator end();
+        const_iterator end() const;
+        const_iterator cend() const;
+        ///@}
+
+        /*
+         * @return A range of the bitmap.
+         */
+        const_bitmap_range bitmap() const;
+
+        /*
+         * @return A range of the values.
+         */
+        const_value_range values() const;
+
+        // Capacity
+
+        /*
+         * @return true if the container is empty, false otherwise.
+         */
+        bool empty() const;
+
+        /*
+         * @return The number of elements in the container.
+         */
+        size_type size() const;
+
+        // TODO: Add reserve, capacity, shrink_to_fit
+
+        // Modifiers
+
+        // TODO: Implement insert, erase, push_back, pop_back, clear, resize, swap
+
+        std::partial_ordering operator<=>(const typed_array<T>& other) const;
+
+        bool operator==(const typed_array<T>& other) const;
+
+    private:
+
+        array_data m_data;
+        layout_type m_layout;
+    };
+
+    // Constructors
+    template <class T, class L>
+        requires is_arrow_base_type<T>
+    typed_array<T, L>::typed_array(array_data data)
+        : m_data(std::move(data))
+        , m_layout(m_data)
+    {
+    }
+
+    // Element access
+
+    template <class T, class L>
+        requires is_arrow_base_type<T>
+    auto typed_array<T, L>::at(size_type i) -> reference
+    {
+        if (i >= size())
+        {
+#if defined(COMPILING_WITH_APPLE_CLANG)
+            throw std::out_of_range(
+                "typed_array::at: index out of range for array of size " + std::to_string(size())
+                + " at index " + std::to_string(i)
+            );
+#else
+            throw std::out_of_range(
+                std::format("typed_array::at: index out of range for array of size {} at index {}", size(), i)
+            );
+#endif
+        }
+        return m_layout[i];
+    }
+
+    template <class T, class L>
+        requires is_arrow_base_type<T>
+    auto typed_array<T, L>::at(size_type i) const -> const_reference
+    {
+        if (i >= size())
+        {
+#if defined(COMPILING_WITH_APPLE_CLANG)
```

Comment:
We might want to make our own formatting function that forwards to std::format or  custom impl based on std::stringstream. Can wait another PR of course.

Reply:
Ok, I will keep the std::to_string implementation and wait for our custom format implemenation. I'm struggle with the GCC support on Windows and it's not the point of this PR.